==7627== Memcheck, a memory error detector
==7627== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==7627== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==7627== Command: ./FlashLogger
==7627== Parent PID: 1875
==7627== 
==7627== Thread 2:
==7627== Conditional jump or move depends on uninitialised value(s)
==7627==    at 0x41865A: LogFile::append(char const*, long long) (LogFile.cpp:34)
==7627==    by 0x419F16: AsynLogger::backgroundFunc() (AsynLogger.cpp:146)
==7627==    by 0x41B5DC: void std::__invoke_impl<void, void (AsynLogger::*)(), AsynLogger*>(std::__invoke_memfun_deref, void (AsynLogger::*&&)(), AsynLogger*&&) (invoke.h:73)
==7627==    by 0x41AA4A: std::__invoke_result<void (AsynLogger::*)(), AsynLogger*>::type std::__invoke<void (AsynLogger::*)(), AsynLogger*>(void (AsynLogger::*&&)(), AsynLogger*&&) (invoke.h:95)
==7627==    by 0x41CF94: decltype (__invoke((_S_declval<0ul>)(), (_S_declval<1ul>)())) std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) (thread:244)
==7627==    by 0x41CF4F: std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> >::operator()() (thread:253)
==7627==    by 0x41CF33: std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> > >::_M_run() (thread:196)
==7627==    by 0x5128BA2: execute_native_thread_routine (thread.cc:80)
==7627==    by 0x4E4E179: start_thread (in /usr/lib64/libpthread-2.28.so)
==7627==    by 0x5A91DC2: clone (in /usr/lib64/libc-2.28.so)
==7627== 
==7627== 
==7627== HEAP SUMMARY:
==7627==     in use at exit: 167,773,009 bytes in 35 blocks
==7627==   total heap usage: 61 allocs, 26 frees, 167,857,978 bytes allocated
==7627== 
==7627== Thread 1:
==7627== 1 bytes in 1 blocks are still reachable in loss record 1 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x41D0C5: Logger::getInstance() (Logger.cpp:41)
==7627==    by 0x41E3DF: func() (AsynLoggerBenchmark.cpp:26)
==7627==    by 0x41EF44: void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) (invoke.h:60)
==7627==    by 0x41EB06: std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) (invoke.h:95)
==7627==    by 0x41F513: decltype (__invoke((_S_declval<0ul>)())) std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) (thread:244)
==7627==    by 0x41F4E9: std::thread::_Invoker<std::tuple<void (*)()> >::operator()() (thread:253)
==7627==    by 0x41F4CD: std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() (thread:196)
==7627==    by 0x5128BA2: execute_native_thread_routine (thread.cc:80)
==7627==    by 0x4E4E179: start_thread (in /usr/lib64/libpthread-2.28.so)
==7627==    by 0x5A91DC2: clone (in /usr/lib64/libc-2.28.so)
==7627== 
==7627== 24 bytes in 1 blocks are indirectly lost in loss record 2 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x41C637: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<AsynLogger::BufferNode*>(AsynLogger::BufferNode*) (shared_ptr_base.h:623)
==7627==    by 0x41BD72: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<AsynLogger::BufferNode*>(AsynLogger::BufferNode*, std::integral_constant<bool, false>) (shared_ptr_base.h:634)
==7627==    by 0x41B27D: std::__shared_ptr<AsynLogger::BufferNode, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<AsynLogger::BufferNode, void>(AsynLogger::BufferNode*) (shared_ptr_base.h:1123)
==7627==    by 0x41A8C8: std::shared_ptr<AsynLogger::BufferNode>::shared_ptr<AsynLogger::BufferNode, void>(AsynLogger::BufferNode*) (shared_ptr.h:139)
==7627==    by 0x419426: AsynLogger::AsynLogger(int, int) (AsynLogger.cpp:15)
==7627==    by 0x41E4DA: main (AsynLoggerBenchmark.cpp:45)
==7627== 
==7627== 24 bytes in 1 blocks are indirectly lost in loss record 3 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x41C637: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<AsynLogger::BufferNode*>(AsynLogger::BufferNode*) (shared_ptr_base.h:623)
==7627==    by 0x41BD72: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<AsynLogger::BufferNode*>(AsynLogger::BufferNode*, std::integral_constant<bool, false>) (shared_ptr_base.h:634)
==7627==    by 0x41B27D: std::__shared_ptr<AsynLogger::BufferNode, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<AsynLogger::BufferNode, void>(AsynLogger::BufferNode*) (shared_ptr_base.h:1123)
==7627==    by 0x41A8C8: std::shared_ptr<AsynLogger::BufferNode>::shared_ptr<AsynLogger::BufferNode, void>(AsynLogger::BufferNode*) (shared_ptr.h:139)
==7627==    by 0x419469: AsynLogger::AsynLogger(int, int) (AsynLogger.cpp:16)
==7627==    by 0x41E4DA: main (AsynLoggerBenchmark.cpp:45)
==7627== 
==7627== 24 bytes in 1 blocks are indirectly lost in loss record 4 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x41C637: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<AsynLogger::BufferNode*>(AsynLogger::BufferNode*) (shared_ptr_base.h:623)
==7627==    by 0x41BD72: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<AsynLogger::BufferNode*>(AsynLogger::BufferNode*, std::integral_constant<bool, false>) (shared_ptr_base.h:634)
==7627==    by 0x41B27D: std::__shared_ptr<AsynLogger::BufferNode, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<AsynLogger::BufferNode, void>(AsynLogger::BufferNode*) (shared_ptr_base.h:1123)
==7627==    by 0x41A8C8: std::shared_ptr<AsynLogger::BufferNode>::shared_ptr<AsynLogger::BufferNode, void>(AsynLogger::BufferNode*) (shared_ptr.h:139)
==7627==    by 0x41993A: AsynLogger::newBufferNode() (AsynLogger.cpp:63)
==7627==    by 0x419DCB: AsynLogger::backgroundFunc() (AsynLogger.cpp:136)
==7627==    by 0x41B5DC: void std::__invoke_impl<void, void (AsynLogger::*)(), AsynLogger*>(std::__invoke_memfun_deref, void (AsynLogger::*&&)(), AsynLogger*&&) (invoke.h:73)
==7627==    by 0x41AA4A: std::__invoke_result<void (AsynLogger::*)(), AsynLogger*>::type std::__invoke<void (AsynLogger::*)(), AsynLogger*>(void (AsynLogger::*&&)(), AsynLogger*&&) (invoke.h:95)
==7627==    by 0x41CF94: decltype (__invoke((_S_declval<0ul>)(), (_S_declval<1ul>)())) std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) (thread:244)
==7627==    by 0x41CF4F: std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> >::operator()() (thread:253)
==7627==    by 0x41CF33: std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> > >::_M_run() (thread:196)
==7627== 
==7627== 40 bytes in 1 blocks are indirectly lost in loss record 5 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x41944F: AsynLogger::AsynLogger(int, int) (AsynLogger.cpp:16)
==7627==    by 0x41E4DA: main (AsynLoggerBenchmark.cpp:45)
==7627== 
==7627== 40 bytes in 1 blocks are indirectly lost in loss record 6 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x419920: AsynLogger::newBufferNode() (AsynLogger.cpp:63)
==7627==    by 0x419DCB: AsynLogger::backgroundFunc() (AsynLogger.cpp:136)
==7627==    by 0x41B5DC: void std::__invoke_impl<void, void (AsynLogger::*)(), AsynLogger*>(std::__invoke_memfun_deref, void (AsynLogger::*&&)(), AsynLogger*&&) (invoke.h:73)
==7627==    by 0x41AA4A: std::__invoke_result<void (AsynLogger::*)(), AsynLogger*>::type std::__invoke<void (AsynLogger::*)(), AsynLogger*>(void (AsynLogger::*&&)(), AsynLogger*&&) (invoke.h:95)
==7627==    by 0x41CF94: decltype (__invoke((_S_declval<0ul>)(), (_S_declval<1ul>)())) std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) (thread:244)
==7627==    by 0x41CF4F: std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> >::operator()() (thread:253)
==7627==    by 0x41CF33: std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> > >::_M_run() (thread:196)
==7627==    by 0x5128BA2: execute_native_thread_routine (thread.cc:80)
==7627==    by 0x4E4E179: start_thread (in /usr/lib64/libpthread-2.28.so)
==7627==    by 0x5A91DC2: clone (in /usr/lib64/libc-2.28.so)
==7627== 
==7627== 72 bytes in 3 blocks are indirectly lost in loss record 7 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x41C637: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<AsynLogger::BufferNode*>(AsynLogger::BufferNode*) (shared_ptr_base.h:623)
==7627==    by 0x41BD72: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<AsynLogger::BufferNode*>(AsynLogger::BufferNode*, std::integral_constant<bool, false>) (shared_ptr_base.h:634)
==7627==    by 0x41B27D: std::__shared_ptr<AsynLogger::BufferNode, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<AsynLogger::BufferNode, void>(AsynLogger::BufferNode*) (shared_ptr_base.h:1123)
==7627==    by 0x41A8C8: std::shared_ptr<AsynLogger::BufferNode>::shared_ptr<AsynLogger::BufferNode, void>(AsynLogger::BufferNode*) (shared_ptr.h:139)
==7627==    by 0x41993A: AsynLogger::newBufferNode() (AsynLogger.cpp:63)
==7627==    by 0x419758: AsynLogger::start() (AsynLogger.cpp:43)
==7627==    by 0x41E518: main (AsynLoggerBenchmark.cpp:46)
==7627== 
==7627== 120 bytes in 3 blocks are indirectly lost in loss record 8 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x419920: AsynLogger::newBufferNode() (AsynLogger.cpp:63)
==7627==    by 0x419758: AsynLogger::start() (AsynLogger.cpp:43)
==7627==    by 0x41E518: main (AsynLoggerBenchmark.cpp:46)
==7627== 
==7627== 144 bytes in 6 blocks are indirectly lost in loss record 9 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x41C637: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<AsynLogger::BufferNode*>(AsynLogger::BufferNode*) (shared_ptr_base.h:623)
==7627==    by 0x41BD72: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<AsynLogger::BufferNode*>(AsynLogger::BufferNode*, std::integral_constant<bool, false>) (shared_ptr_base.h:634)
==7627==    by 0x41B27D: std::__shared_ptr<AsynLogger::BufferNode, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<AsynLogger::BufferNode, void>(AsynLogger::BufferNode*) (shared_ptr_base.h:1123)
==7627==    by 0x41A8C8: std::shared_ptr<AsynLogger::BufferNode>::shared_ptr<AsynLogger::BufferNode, void>(AsynLogger::BufferNode*) (shared_ptr.h:139)
==7627==    by 0x41993A: AsynLogger::newBufferNode() (AsynLogger.cpp:63)
==7627==    by 0x419A97: AsynLogger::append(char const*, unsigned long) (AsynLogger.cpp:91)
==7627==    by 0x41E440: output(char const*, unsigned long) (AsynLoggerBenchmark.cpp:34)
==7627==    by 0x41DF64: std::_Function_handler<void (char const*, unsigned long), void (*)(char const*, unsigned long)>::_M_invoke(std::_Any_data const&, char const*&&, unsigned long&&) (std_function.h:297)
==7627==    by 0x41DEA2: std::function<void (char const*, unsigned long)>::operator()(char const*, unsigned long) const (std_function.h:687)
==7627==    by 0x41D512: Logger::append(char const*, Logger::LogLevel, char const*, char const*, ...) (Logger.cpp:133)
==7627==    by 0x41E407: func() (AsynLoggerBenchmark.cpp:26)
==7627== 
==7627== 240 bytes in 6 blocks are indirectly lost in loss record 10 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x419920: AsynLogger::newBufferNode() (AsynLogger.cpp:63)
==7627==    by 0x419A97: AsynLogger::append(char const*, unsigned long) (AsynLogger.cpp:91)
==7627==    by 0x41E440: output(char const*, unsigned long) (AsynLoggerBenchmark.cpp:34)
==7627==    by 0x41DF64: std::_Function_handler<void (char const*, unsigned long), void (*)(char const*, unsigned long)>::_M_invoke(std::_Any_data const&, char const*&&, unsigned long&&) (std_function.h:297)
==7627==    by 0x41DEA2: std::function<void (char const*, unsigned long)>::operator()(char const*, unsigned long) const (std_function.h:687)
==7627==    by 0x41D512: Logger::append(char const*, Logger::LogLevel, char const*, char const*, ...) (Logger.cpp:133)
==7627==    by 0x41E407: func() (AsynLoggerBenchmark.cpp:26)
==7627==    by 0x41EF44: void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) (invoke.h:60)
==7627==    by 0x41EB06: std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) (invoke.h:95)
==7627==    by 0x41F513: decltype (__invoke((_S_declval<0ul>)())) std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) (thread:244)
==7627==    by 0x41F4E9: std::thread::_Invoker<std::tuple<void (*)()> >::operator()() (thread:253)
==7627== 
==7627== 16,777,224 bytes in 1 blocks are possibly lost in loss record 11 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x419944: AsynLogger::newBufferNode() (AsynLogger.cpp:64)
==7627==    by 0x419DCB: AsynLogger::backgroundFunc() (AsynLogger.cpp:136)
==7627==    by 0x41B5DC: void std::__invoke_impl<void, void (AsynLogger::*)(), AsynLogger*>(std::__invoke_memfun_deref, void (AsynLogger::*&&)(), AsynLogger*&&) (invoke.h:73)
==7627==    by 0x41AA4A: std::__invoke_result<void (AsynLogger::*)(), AsynLogger*>::type std::__invoke<void (AsynLogger::*)(), AsynLogger*>(void (AsynLogger::*&&)(), AsynLogger*&&) (invoke.h:95)
==7627==    by 0x41CF94: decltype (__invoke((_S_declval<0ul>)(), (_S_declval<1ul>)())) std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> >::_M_invoke<0ul, 1ul>(std::_Index_tuple<0ul, 1ul>) (thread:244)
==7627==    by 0x41CF4F: std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> >::operator()() (thread:253)
==7627==    by 0x41CF33: std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (AsynLogger::*)(), AsynLogger*> > >::_M_run() (thread:196)
==7627==    by 0x5128BA2: execute_native_thread_routine (thread.cc:80)
==7627==    by 0x4E4E179: start_thread (in /usr/lib64/libpthread-2.28.so)
==7627==    by 0x5A91DC2: clone (in /usr/lib64/libc-2.28.so)
==7627== 
==7627== 50,331,672 bytes in 3 blocks are possibly lost in loss record 12 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x419944: AsynLogger::newBufferNode() (AsynLogger.cpp:64)
==7627==    by 0x419758: AsynLogger::start() (AsynLogger.cpp:43)
==7627==    by 0x41E518: main (AsynLoggerBenchmark.cpp:46)
==7627== 
==7627== 50,331,672 bytes in 3 blocks are possibly lost in loss record 13 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x419944: AsynLogger::newBufferNode() (AsynLogger.cpp:64)
==7627==    by 0x419A97: AsynLogger::append(char const*, unsigned long) (AsynLogger.cpp:91)
==7627==    by 0x41E440: output(char const*, unsigned long) (AsynLoggerBenchmark.cpp:34)
==7627==    by 0x41DF64: std::_Function_handler<void (char const*, unsigned long), void (*)(char const*, unsigned long)>::_M_invoke(std::_Any_data const&, char const*&&, unsigned long&&) (std_function.h:297)
==7627==    by 0x41DEA2: std::function<void (char const*, unsigned long)>::operator()(char const*, unsigned long) const (std_function.h:687)
==7627==    by 0x41D512: Logger::append(char const*, Logger::LogLevel, char const*, char const*, ...) (Logger.cpp:133)
==7627==    by 0x41E407: func() (AsynLoggerBenchmark.cpp:26)
==7627==    by 0x41EF44: void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) (invoke.h:60)
==7627==    by 0x41EB06: std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) (invoke.h:95)
==7627==    by 0x41F513: decltype (__invoke((_S_declval<0ul>)())) std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) (thread:244)
==7627==    by 0x41F4E9: std::thread::_Invoker<std::tuple<void (*)()> >::operator()() (thread:253)
==7627== 
==7627== 50,331,672 bytes in 3 blocks are indirectly lost in loss record 14 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x419944: AsynLogger::newBufferNode() (AsynLogger.cpp:64)
==7627==    by 0x419A97: AsynLogger::append(char const*, unsigned long) (AsynLogger.cpp:91)
==7627==    by 0x41E440: output(char const*, unsigned long) (AsynLoggerBenchmark.cpp:34)
==7627==    by 0x41DF64: std::_Function_handler<void (char const*, unsigned long), void (*)(char const*, unsigned long)>::_M_invoke(std::_Any_data const&, char const*&&, unsigned long&&) (std_function.h:297)
==7627==    by 0x41DEA2: std::function<void (char const*, unsigned long)>::operator()(char const*, unsigned long) const (std_function.h:687)
==7627==    by 0x41D512: Logger::append(char const*, Logger::LogLevel, char const*, char const*, ...) (Logger.cpp:133)
==7627==    by 0x41E407: func() (AsynLoggerBenchmark.cpp:26)
==7627==    by 0x41EF44: void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) (invoke.h:60)
==7627==    by 0x41EB06: std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) (invoke.h:95)
==7627==    by 0x41F513: decltype (__invoke((_S_declval<0ul>)())) std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) (thread:244)
==7627==    by 0x41F4E9: std::thread::_Invoker<std::tuple<void (*)()> >::operator()() (thread:253)
==7627== 
==7627== 50,332,440 (40 direct, 50,332,400 indirect) bytes in 1 blocks are definitely lost in loss record 15 of 15
==7627==    at 0x4C3676F: operator new(unsigned long) (vg_replace_malloc.c:422)
==7627==    by 0x41940C: AsynLogger::AsynLogger(int, int) (AsynLogger.cpp:15)
==7627==    by 0x41E4DA: main (AsynLoggerBenchmark.cpp:45)
==7627== 
==7627== LEAK SUMMARY:
==7627==    definitely lost: 40 bytes in 1 blocks
==7627==    indirectly lost: 50,332,400 bytes in 26 blocks
==7627==      possibly lost: 117,440,568 bytes in 7 blocks
==7627==    still reachable: 1 bytes in 1 blocks
==7627==         suppressed: 0 bytes in 0 blocks
==7627== 
==7627== Use --track-origins=yes to see where uninitialised values come from
==7627== For lists of detected and suppressed errors, rerun with: -s
==7627== ERROR SUMMARY: 13 errors from 5 contexts (suppressed: 0 from 0)
